crates/ra_hir_def/src/nameres/collector.rs

```diff
@@ -482,6 +496,77 @@ where
         res
     }
 
+    fn expand_eager_macros(
+        &mut self,
+        eagers: Vec<(CrateModuleId, AstId<ast::MacroCall>, MacroDefId, HirFileId)>,
+    ) -> Option<()> {
+        for (module_id, ast_id, def, file_id) in eagers.into_iter() {
+            let resolve_file_id = &|relative_path: &str| {
+                let relative_path = RelativePath::new(relative_path);
+                self.db.resolve_relative_path(file_id.original_file(self.db), relative_path)
+            };
+            let resolve_file_content = &|relative_path: &str| {
+                let content = self.db.file_text(resolve_file_id(relative_path)?);
+                Some((*content).clone())
+            };
+
+            let resolve_macro = &|path: ast::Path| {
+                let path = Path::from_src(path.clone(), &Hygiene::new(self.db, file_id))?;
+
+                // legacy scope
+                if let Some(macro_def) = path
+                    .as_ident()
+                    .and_then(|name| self.def_map[module_id].scope.get_legacy_macro(&name))
+                {
+                    return Some(MacroDefIdWithAst::from_def(self.db, macro_def));
+                }
+
+                let mut path = path.clone();
+                if path.is_ident() {
+                    path.kind = PathKind::Self_;
+                }
+
+                // Resolve in module scope, expand during name resolution.
+                let resolved_res = self.def_map.resolve_path_fp_with_macro(
+                    self.db,
+                    ResolveMode::Other,
+                    module_id,
+                    &path,
+                );
+
+                resolved_res
+                    .resolved_def
+                    .get_macros()
+                    .map(|it| MacroDefIdWithAst::from_def(self.db, it))
+            };
+
+            let eager_res = eager::expand_eager_macro(
+                ast_id.to_node(self.db),
```

Comment:
Hm, this will adds a dependecy edge from CrateDefMap to *syntax* of a macro, and that breaks incrementality.

Reply:
Um... what should I do instead ?

Reply:
Hm, let's change 

```rust
fn macro_arg(&self, id: MacroCallId) -> Option<Arc<(tt::Subtree, mbe::TokenMap)>>;
```

to

```rust
fn macro_arg(&self, id: AstId<ast::MacroCall>) -> Option<Arc<(tt::Subtree, mbe::TokenMap)>>;
```

I think this should work... If it does, it makes sense to send a separate PR for that!